### PR TITLE
Adding endpoint for order

### PR DESCRIPTION
**Pull request by simonfabcic**
[What?]
Adding endpoint for order.

[Why?]
Order can be changed from frontend.
Possible changes:
- add product
- remove product
- change the quantity of existing product

[How?]
- Create a serializer for `Order` and for `OrderItem`.
In `OrderSerializer` I have to override the `create` and `update` methods because I have nested serializer.
- I created a custom permission class `CanModifyOrViewOrder` in which I check if the user has permission for the current method.
- Added URLs for `order` endpoint
- Added view `OrderViewSet`

[Testing?]
- User with `view_order`:
    - can view order
    - can't create order
    - can't edit order
- User with `change_order`:
    - can view order
    - can create order
    - can edit order
- Unauthenticated user:
    - can't view order
    - can't create order
    - can't edit order

### CODE REVIEW

File: backend/shop/models.py [hunk 1]
```diff
@@ -15,14 +15,19 @@ class Order(models.Model):
 
 
 class Product(models.Model):
-    image = models.ImageField(blank=True, default="", upload_to="images/products")
+    image = models.ImageField(null=True, default="", upload_to="images/products")
     title = models.CharField(max_length=255)
-    description = models.TextField(blank=True, null=True)
+    description = models.TextField(null=True)
```
Comment (jurecuhalev) [inline backend/shop/models.py:20]: you want text based fields to be `blank=True, default=""`

File: backend/shop/serializers.py [hunk 2]
```diff
@@ -51,3 +52,53 @@ class ProductSerializer(ModelSerializer):
...
+        # Delete any OrderItems that are no longer in the update data
+        for order_item in instance.order_items.all():
+            if order_item.product.id not in order_item_ids:
+                order_item.delete()
+
+        # Update or create OrderItems
+        for order_item_data in order_items_data:
+            try:
+                order_item = OrderItem.objects.get(
+                    # order=instance, product=order_item_data["product"]
```
Comment (jurecuhalev) [inline backend/shop/serializers.py:95]: remove debug comment

File: backend/shop/serializers.py [hunk 3]
```diff
@@ -51,3 +52,53 @@ class ProductSerializer(ModelSerializer):
...
+                    order=instance,
+                    product_id=order_item_data["product"].id,
+                )
+                order_item.quantity = order_item_data["quantity"]
+                order_item.save()
+            except:
```
Comment (jurecuhalev) [inline backend/shop/serializers.py:101]: no blank excepts (please add it to your CI checks)

What are you trying to catch from?

File: backend/shop/urls.py [hunk 4]
```diff
@@ -10,7 +14,9 @@
     path("role/new/", views.role_create, name="role_create"),
     path("permission/", views.permission_get, name="permission_get"),
     path("permission-user/", views.permission_user_get, name="permission_user"),
-    path("product/", views.product_get, name="product_get_all"),
+    path("product/", views.product_get, name="product_get_many"),
```
Comment (jurecuhalev) [inline backend/shop/urls.py:17]: To me get many doesn't tell me anything - it's confusing (why many, why not all, how do I tell how many)? Instead think of a model of -> "product_list" and then you can have "product_list" with a filter applied. More of a pipeline approach https://en.wikipedia.org/wiki/Pipeline_(computing)

File: backend/shop/views.py [hunk 5]
```diff
@@ -7,20 +7,20 @@
 from django.core.paginator import Paginator
 from django.db import IntegrityError, connection
 from django.shortcuts import render
-from rest_framework import status
+from rest_framework import status, viewsets
 from rest_framework.decorators import api_view, permission_classes
-from rest_framework.permissions import IsAuthenticated
+from rest_framework.permissions import IsAuthenticated, IsAuthenticatedOrReadOnly
```
Comment (jurecuhalev) [inline backend/shop/views.py:12]: Unused import. Please fix your CI so that if will if you have unused imports in your PY files.

**Review by jurecuhalev — CHANGES_REQUESTED** [review]
Please fix your CI so that it works. CI Tests shouldn't pass for this PR.

File: backend/shop/models.py (hunk 1 above)
Comment (simonfabcic) [reply backend/shop/models.py:20]: Did you think:
```
description = models.TextField(null=True, blank=True, default="")
```
or 
```
description = models.TextField(blank=True, default="")
```

Comment (jurecuhalev) [reply backend/shop/models.py:20]: second one, do you __ever__ want your TextField to be null (`None`)?

**Comment by simonfabcic** [comment]
CI will be updated in one of the next PR.

File: backend/shop/models.py [hunk 6]
```diff
@@ -15,14 +15,19 @@ class Order(models.Model):
...
 class Product(models.Model):
-    image = models.ImageField(blank=True, default="", upload_to="images/products")
+    image = models.ImageField(null=True, default="", upload_to="images/products")
     title = models.CharField(max_length=255)
-    description = models.TextField(blank=True, null=True)
+    description = models.TextField(blank=True, default="")
     price = models.DecimalField(max_digits=10, decimal_places=2)
 
+    def __str__(self) -> str:
+        return self.title + ", " + str(self.price) + " $"
```
Comment (jurecuhalev) [inline backend/shop/models.py:24]: use f string

File: backend/shop/tests/test_views.py [hunk 7]
```diff
@@ -504,7 +505,7 @@ def setUp(self):
         perm_change_product = Permission.objects.get(codename="change_product")
         self.user_product_editor.user_permissions.add(perm_change_product)
 
-        self.url_all = reverse("product_get_all")
+        self.url_all = reverse("product_list")
```
Comment (jurecuhalev) [inline backend/shop/tests/test_views.py:508]: be consistent, `        self.url_list = reverse("order-list")` one is with `-` and another with `_`.
Recommendation: use `-`

**Review by jurecuhalev — CHANGES_REQUESTED** [review]
See comments

**Pull request merged**